client/finality-grandpa/src/environment.rs

```diff
@@ -815,6 +815,59 @@ where
 		Ok(())
 	}
 
+	fn concluded(
+		&self,
+		round: RoundNumber,
+		state: RoundState<Block::Hash, NumberFor<Block>>,
+		_base: (Block::Hash, NumberFor<Block>),
+		historical_votes: &HistoricalVotes<Block>,
+	) -> Result<(), Self::Error> {
+		debug!(
+			target: "afg", "Voter {} concluded round {} in set {}. Estimate = {:?}, Finalized in round = {:?}",
+			self.config.name(),
+			round,
+			self.set_id,
+			state.estimate.as_ref().map(|e| e.1),
+			state.finalized.as_ref().map(|e| e.1),
+		);
+
+		self.update_voter_set_state(|voter_set_state| {
+			// NOTE: we don't use `with_current_round` here, it is possible that
+			// we are not currently tracking this round if it is a round we
+			// caught up to.
+			let (completed_rounds, current_rounds) =
+				if let VoterSetState::Live { completed_rounds, current_rounds } = voter_set_state {
+					(completed_rounds, current_rounds)
+				} else {
+					let msg = "Voter acting while in paused state.";
+					return Err(Error::Safety(msg.to_string()));
+				};
+
+			let mut completed_rounds = completed_rounds.clone();
+
+			if let Some(already_completed) = completed_rounds.rounds
+				.iter_mut().find(|r| r.number == round)
+			{
+				let n_existing_votes = already_completed.votes.len();
+				already_completed.votes.extend(
+					historical_votes.seen().iter().skip(n_existing_votes).cloned()
```

Comment:
Maybe add a comment here about the contract that `historical_votes` seen in completed must be a prefix of `historical_votes` in concluded.